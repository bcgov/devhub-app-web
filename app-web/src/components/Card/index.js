/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

// common card layout components
import styled from '@emotion/styled';
import { RESOURCE_TYPES, TOPICS, SEARCH_RESOURCE_TYPES, JOURNEY } from '../../constants/ui';
import Dotdotdot from 'react-dotdotdot';
import Link from '../UI/Link/Link';
import { EventLogoWrapper, DateLI, EventDetailsContainer } from '../Event';
export { BaseCard, Card } from './Card';

export const Container = styled.article`
  width: 265px;
  height: 238.5px;
  margin: 10px 0;
  border: 1px solid #ccc;
  border-top: transparent;
  border-radius: 2px;
  box-sizing: border-box;
  transition: transform 0.25s ease-out;
  box-shadow: 0 1px 2px 1px #00000026;
  display: flex;
  flex-flow: column nowrap;
  overflow: hidden;
  :hover {
    box-shadow: 0 2px 2px 1px #00000026;
    transform: translateY(-2px);
  }
`;

export const DecorativeBar = styled.div`
  background-color: ${props => props.theme.colors[props.color]};
  flex: 0 0 10px;
  min-height: 10px;
`;

export const Title = styled(Dotdotdot)`
  font-size: 20px;
  color: ${({ theme }) => theme.colors.grey};
  line-height: 1.3;
  margin-bottom: 0;
  flex: 0 0 auto;
  text-transform: capitalize;
  word-break: break-word;
`;

export const Description = styled(Dotdotdot)`
  line-height: 1.4;
  font-size: 14px;
  flex: 0 0 auto;
  margin-bottom: 0;
  margin-top: 4px;
  text-transform: none;
`;

export const LinkWrapper = styled(Link)`
  text-decoration: none;
  color: initial;
  :hover {
    text-decoration: none;
    color: initial;
    cursor: pointer;
  }
`;

/**
 * used to predictably lay out titles/ descriptions/images in a flex box within a card
 */
export const CardBody = styled.div`
  padding: 6px 10px;
  height: 100%;
  overflow: auto;
  height: 100%;
  display: flex;
  flex: 1 1 auto;
  flex-flow: column nowrap;
`;

// combination of card image and wrapper make images response correctly
// to the flex container. The card image wrapper can grow/shrink automatically
// the object-fit scaledown maintains aspect ratio for any type of image
export const Image = styled.img`
  width: 100%;
  height: auto;
  object-fit: scale-down;
  max-height: 100%;
  margin-bottom: 0;
`;

export const ImageWrapper = styled.div`
  flex: 1 1 auto;
  align-self: center;
  overflow: hidden;
  padding-top: 10px;
  text-align: center;
  display: flex;
`;

export const EventInfoDiv = styled.div`
  > li {
    color: #494949;
    margin-bottom: 0px;
  }
`;

export const EventDate = styled(DateLI)`
  border-radius: 0px;
  border: 0px;
  margin: 0;
  > span {
    margin-top: -15px;
    color: #494949;
    font-weight: 1000;
  }
  > small {
    bottom: 0.1em;
  }
`;

export const EventContainer = styled(EventDetailsContainer)`
  height: 64px;
  font-size: 15px;
  margin-left: -10px;
  background: rgba(229, 229, 229, 0.8);
  width: 100%;
  > li {
    margin-bottom: 0px;
  }
`;

// special cases to handle presentation of the eventbrite and meetup logo inside of a card
// please note that although meetups is no longer being implemented (since the api disabled access token authenticated)
// it may be reactivated at one point, for this reason these components are being left behind
export const EventbriteImageWrapper = styled(EventLogoWrapper)`
  margin-top: 0px;
  img {
    margin-left: -8px;
    padding-left: 0px;
  }
`;

export const MeetupImageWrapper = styled(EventLogoWrapper)`
  margin: -1px 0px 5px 0px;
  img {
    padding-left: 0px;
  }
`;

export const CONTENT = {
  byResourceType: {
    [RESOURCE_TYPES.COMPONENTS]: {
      id: 'RESOURCE_TYPES.COMPONENTS',
      text: 'Component',
    },
    [RESOURCE_TYPES.DOCUMENTATION]: {
      id: 'RESOURCE_TYPES.DOCUMENTATION',
      text: 'Documentation',
    },
    [RESOURCE_TYPES.SELF_SERVICE_TOOLS]: {
      id: 'RESOURCE_TYPES.SELF_SERVICE_TOOLS',
      text: 'Tool',
    },
    [RESOURCE_TYPES.PEOPLE]: {
      id: 'RESOURCE_TYPES.PEOPLE',
      text: 'Contact',
    },
    [RESOURCE_TYPES.REPOSITORIES]: {
      id: 'RESOURCE_TYPES.REPOSITORIES',
      text: 'Repository',
    },
    [TOPICS]: {
      id: 'RESOURCE_TYPES.TOPICS',
      text: 'Topic',
    },
    [RESOURCE_TYPES.EVENTS]: {
      id: 'RESOURCE_TYPES.EVENTS',
      text: 'Event',
    },
    [SEARCH_RESOURCE_TYPES.GITHUB_ISSUE]: {
      id: 'SEARCH_RESOURCE_TYPES.GITHUB_ISSUE',
      text: 'Github Issue',
    },
    [JOURNEY]: {
      id: 'RESOURCE_TYPES.JOURNEY',
      text: 'Journey',
    },
  },
};
