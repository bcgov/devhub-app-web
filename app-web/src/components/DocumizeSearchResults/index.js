import React from 'react';
import styled from '@emotion/styled';
import rocketChatSvg from '../../assets/images/rocketchat_logo.svg';
import { RocketChatItem } from './RocketChatItem';
import { Title, LinkContainer } from '../Home';
import { ChevronLink } from '../UI/Link';
import { SEARCH_SOURCE_CONFIG } from '../../constants/search';
const Container = styled.div`
  padding: 0;
  margin: 0;
  list-style: none;
  max-width: 1100px;
  margin: 20px auto;
  img {
    width: 40px;
    margin-bottom: 0;
    margin-left: 3px;
  }
`;

export const RocketChatResults = ({ results }) => (
  <Container>
    <Title id={'rocketChat'}>
      From Rocket.Chat <img src={rocketChatSvg} alt="rocket chat logo" />
    </Title>
    {results.length > SEARCH_SOURCE_CONFIG.rocketchat.maxResults && (
      <small>
        Showing {SEARCH_SOURCE_CONFIG.rocketchat.maxResults} of {results.length}
      </small>
    )}
    {results.slice(0, SEARCH_SOURCE_CONFIG.rocketchat.maxResults).map(r => (
      <RocketChatItem {...r} key={r.id} data-testid={r.id} />
    ))}
    <LinkContainer>
      <ChevronLink to="https://chat.pathfinder.gov.bc.ca">Go to Rocket.Chat</ChevronLink>
    </LinkContainer>
  </Container>
);
