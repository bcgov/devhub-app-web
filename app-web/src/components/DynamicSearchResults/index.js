import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { Title, LinkContainer } from '../Home';
import { ChevronLink } from '../UI/Link';
import { SEARCH_SOURCE_CONFIG, SEARCH_SOURCES } from '../../constants/search';

import documizeLogo from '../../assets/images/documize_logo.png';
import rocketchatLogo from '../../assets/images/rocketchat_logo.svg';

const Container = styled.div`
  padding: 0;
  margin: 0;
  list-style: none;
  max-width: 1100px;
  margin: 20px auto;
  img {
    width: 40px;
    margin-bottom: 0;
    margin-left: 3px;
  }
`;

export const DynamicSearchResults = ({ results, sourceType, link, renderItem, ...rest }) => {
  const contentMapping = {
    [SEARCH_SOURCES.documize]: {
      name: 'From Documize',
      logo: documizeLogo,
      id: 'documize',
    },
    [SEARCH_SOURCES.rocketchat]: {
      name: 'From Rocket.Chat',
      logo: rocketchatLogo,
      id: 'rocketChat',
    },
  };

  const content = contentMapping[sourceType];
  const settings = SEARCH_SOURCE_CONFIG[sourceType]
    ? SEARCH_SOURCE_CONFIG[sourceType]
    : SEARCH_SOURCE_CONFIG.default;

  return (
    <Container {...rest}>
      <Title id={content.id}>
        {content.name} <img src={content.logo} alt={content.name} />
      </Title>

      {results.length > settings.maxResults && (
        <small>
          Showing {settings.maxResults} of {results.length}
        </small>
      )}
      {results.slice(0, settings.maxResults).map(r => renderItem(r))}
      <LinkContainer>
        <ChevronLink to={link.to}>{link.text}</ChevronLink>
      </LinkContainer>
    </Container>
  );
};

DynamicSearchResults.propTypes = {
  link: PropTypes.shape({
    text: PropTypes.string.isRequired,
    to: PropTypes.string.isRequired,
  }),
  renderItem: PropTypes.func,
  results: PropTypes.array,
  sourceType: PropTypes.string,
};
