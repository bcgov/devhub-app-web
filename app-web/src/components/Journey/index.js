import React from 'react';
import { Link } from '../UI/Link';
import { ResourceTypeIcon } from '../UI/ResourceTypeIcon';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import isFunction from 'lodash/isFunction';
import { EMOTION_BOOTSTRAP_BREAKPOINTS } from '../../constants/designTokens';

const Container = styled(Link)`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  text-transform: capitalize;
  text-align: center;
`;

const stationVariants = ['up', 'down'];

const Circle = styled.span`
  flex-shrink: 0;
  cursor: pointer;
  position: relative;
  display: inline-block;
  transition: all 0.15s ease-out;
  border-radius: 50%;
  background-color: ${({ color, theme }) =>
    color && theme.colors[color] ? theme.colors[color] : theme.colors.blue};
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  ::after {
    position: absolute;
    content: '';
    width: 90%;
    height: 90%;
    border-radius: 50%;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
  }
  width: ${({ size }) => 0.65 * size}px;
  height: ${({ size }) => 0.65 * size}px;
  ${EMOTION_BOOTSTRAP_BREAKPOINTS.md} {
    width: ${({ size }) => 0.8 * size}px;
    height: ${({ size }) => 0.8 * size}px;
  }
  ${EMOTION_BOOTSTRAP_BREAKPOINTS.lg} {
    width: ${({ size }) => size}px;
    height: ${({ size }) => size}px;
  }
`;

const Name = styled.span`
  position: absolute;
  top: ${({ variant }) => (variant === 'up' ? `initial` : `50%`)};
  left: 50%;
  bottom: ${({ variant }) => (variant === 'down' ? `initial` : `50%`)};
  transform: ${({ variant, size }) =>
      variant === 'up' ? `translateY(-${size - 5}px)` : `translateY(${size - 5}px)`}
    translateX(-50%);
  color: ${({ theme }) => theme.colors.darkgrey};
  font-size: ${({ fontSize }) => 0.8 * fontSize}px;
  width: 75px;
  ${EMOTION_BOOTSTRAP_BREAKPOINTS.lg} {
    width: 150px;
    font-size: ${({ fontSize }) => fontSize}px;
  }
`;

const AuxillaryField = styled.span`
  position: absolute;
  top: ${({ variant }) => (variant === 'down' ? `initial` : `50%`)};
  left: 50%;
  bottom: ${({ variant }) => (variant === 'up' ? `initial` : `50%`)};
  transform: ${({ variant, size }) =>
      variant === 'down' ? `translateY(-${size - 5}px)` : `translateY(${size - 5}px)`}
    translateX(-50%);
`;

export const Station = ({ name, variant, color, size = 25, render, ...rest }) => {
  const fontSize = name.length > 24 ? 14 : 16;

  return (
    <Container {...rest}>
      <Circle color={color} size={size} />
      <Name variant={variant} fontSize={fontSize} size={size}>
        {isFunction(name) ? name() : name}
      </Name>{' '}
      {render && (
        <AuxillaryField variant={variant} size={size}>
          {render()}
        </AuxillaryField>
      )}
    </Container>
  );
};

Station.propTypes = {
  name: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  variant: PropTypes.oneOf(stationVariants),
  color: PropTypes.string,
  size: PropTypes.number,
  render: PropTypes.func,
  to: PropTypes.string,
};

Station.defaultProps = {
  variant: 'up',
};

const Line = styled.div`
  height: ${({ size }) => 0.35 * size}px;
  ${EMOTION_BOOTSTRAP_BREAKPOINTS.md} {
    height: ${({ size }) => 0.4 * size}px;
  }
  ${EMOTION_BOOTSTRAP_BREAKPOINTS.lg} {
    height: ${({ size }) => 0.5 * size}px;
  }
  width: calc(100% - ${({ size }) => size}px);
  position: absolute;
  top: 50%;
  transform: translateY(-50%) translateX(${({ size }) => size * 0.5}px);
  background-color: ${({ color, theme }) =>
    color && theme.colors[color] ? theme.colors[color] : theme.colors.blue};
`;

const Subway = styled.div`
  display: flex;
  position: relative;
  align-items: center;
  padding: 100px 0;
  justify-content: space-between;
`;

export const SubwayLine = ({ stops, color, size = 25 }) => (
  <Subway>
    <Line color={color} size={size} />
    {stops.map((stop, index) => {
      const variant = index % 2 === 0 ? 'up' : 'down';
      const extraProps = {};
      if (stop.connections) {
        extraProps.render = () => (
          <JunctionList
            variant={variant === 'up' ? 'down' : 'up'}
            style={{ transform: 'translateX(calc(50% - 2px))' }}
            links={stop.connections.map(({ name, path, resourceType }) => ({
              name,
              to: path,
              resourceType,
            }))}
          />
        );
      }

      return (
        <Station
          key={stop.name}
          name={stop.name}
          size={size}
          variant={variant}
          color={color}
          href={stop.to}
          {...extraProps}
        />
      );
    })}
  </Subway>
);

SubwayLine.propTypes = {
  stops: PropTypes.arrayOf(PropTypes.shape({ name: PropTypes.string })),
  color: PropTypes.string,
  size: PropTypes.number,
};

const JunctionContainer = styled.ul`
  margin: 5px 0;
  border-left: 2px solid ${({ theme }) => theme.colors.darkgrey};
  color: ${({ theme }) => theme.colors.darkgrey};
  font-size: 12px;
  font-weight: 400;
  border-top-left-radius: ${({ variant }) => (variant === 'up' ? '4px' : 0)};
  border-bottom-left-radius: ${({ variant }) => (variant === 'down' ? '4px' : 0)};
  padding: ${({ variant }) => (variant === 'down' ? '1.65em 0 0 4px' : '0 0 1.65em 4px')};
  list-style: none;
  li {
    margin: 6px 0;

    transform: ${({ variant }) => (variant === 'up' ? 'translateY(-0.5em)' : 'translateY(0.5em)')};
    display: flex;
    align-items: flex-end;
    justify-content: flex-start;
    text-align: left;
    a {
      max-width: 150px;
      flex: 1 0 auto;
      color: ${({ theme }) => theme.colors.lightblue};
    }
  }
`;

export const JunctionList = ({ links, variant, ...rest }) => {
  return (
    <JunctionContainer variant={variant} {...rest}>
      {links.map(l => (
        <li key={l.to}>
          <span>
            â”€&nbsp;
            <ResourceTypeIcon type={l.resourceType} style={{ marginRight: '4px' }} />
          </span>
          <Link to={l.to}>{l.name}</Link>
        </li>
      ))}
    </JunctionContainer>
  );
};

JunctionList.propTypes = {
  links: PropTypes.arrayOf(PropTypes.shape({ name: PropTypes.string, to: PropTypes.string })),
  variant: PropTypes.oneOf(stationVariants),
};

JunctionList.defaultProps = {
  variant: 'up',
};
