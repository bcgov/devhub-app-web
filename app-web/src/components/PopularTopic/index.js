import React from 'react';
import PropTypes from 'prop-types';
import { POPULAR_TOPIC_CONFIGURATION, DYNAMIC_TOPIC_PATHS } from '../../constants/ui';
import Topic from '../Topic';

const PopularTopic = ({ nodes }) =>
  nodes.length > 0 ? (
    <Topic
      resources={nodes.map(n => ({
        ...n,
        path: `/topic/${DYNAMIC_TOPIC_PATHS.popular}/${n.fields.slug}`,
      }))}
      title={POPULAR_TOPIC_CONFIGURATION.name}
      description={POPULAR_TOPIC_CONFIGURATION.description}
      link={{ to: `/topic/${DYNAMIC_TOPIC_PATHS.popular}`, text: 'Popular' }}
    />
  ) : null;

PopularTopic.propTypes = {
  nodes: PropTypes.arrayOf(
    PropTypes.shape({
      path: PropTypes.string,
      fields: PropTypes.shape({
        title: PropTypes.string,
        description: PropTypes.string,
        image: PropTypes.string,
        slug: PropTypes.string,
        labels: PropTypes.array,
        tags: PropTypes.array,
        pagePaths: PropTypes.array,
      }),
    }),
  ),
};
export default PopularTopic;
