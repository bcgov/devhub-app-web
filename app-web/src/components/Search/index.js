/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/
import React, { Component } from 'react';
import styled from '@emotion/styled';
import { SEARCH } from '../../messages';
import { ARIA_LABEL_SEARCH_BUTTON, ARIA_LABEL_SEARCH_INPUT } from '../../constants/ariaLabels';
import PropTypes from 'prop-types';
import Button from '../UI/Button/Button';
import Input from '../UI/Input/Input';

const Container = styled.div`
  display: flex;
  align-items: center;
  > input {
    margin: 0 4px;
    padding: 4px 5px;
    flex-basis: 250px;
    flex-grow: 1;
    transition: all 0.25s;
  }
  > button {
    padding: 4px 15px;
    margin: 4px;
  }
`;

export class Search extends Component {
  state = {
    touched: false,
    terms: '',
  };

  handleEnter = e => {
    if (e.key === 'Enter' && this.props.searchOnEnter) {
      this.search();
    }
  };

  handleKeyUp = e => {
    // if enter was pressed
    const terms = e.target.value;
    if (terms === 0 && this.state.touched) {
      this.props.onSearchClear();
    }
    if (e.key !== 'Enter') {
      this.setState({ terms });
    }
  };

  search = () => {
    this.props.onSearch(this.state.terms);
    this.setState({ terms: '' });
  };

  render() {
    const { inputConfig, ...rest } = this.props;
    // rest param is mostly used to pass custom styling to the container if necessaary
    return (
      <Container {...rest}>
        <Input
          type="text"
          aria-label={ARIA_LABEL_SEARCH_INPUT}
          value={this.state.terms}
          onChange={this.handleKeyUp}
          onKeyPress={this.handleEnter}
          onFocus={() => this.setState({ touched: true })}
          {...inputConfig}
        />
        <Button type="primary" aria-label={ARIA_LABEL_SEARCH_BUTTON} clicked={() => this.search()}>
          {SEARCH.button.defaultMessage}
        </Button>
      </Container>
    );
  }
}

Search.propTypes = {
  onSearch: PropTypes.func.isRequired,
  onSearchClear: PropTypes.func,
  searchOnEnter: PropTypes.bool,
  inputConfig: PropTypes.object,
};

Search.defaultProps = {
  onSearchClear: () => undefined,
  searchOnEnter: false,
  inputConfig: {},
};

export default Search;
