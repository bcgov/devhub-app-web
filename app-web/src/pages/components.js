import React, { PureComponent } from 'react';
import { graphql } from 'gatsby';
import { createStructuredSelector } from 'reselect';
import queryString from 'query-string';
import { connect } from 'react-redux';
import { RESOURCE_TYPES } from '../constants/ui';
import { flattenGatsbyGraphQL } from '../utils//dataHelpers';
import * as actions from '../store/actions';

import styles from './index.module.css';
import { COMPONENTS } from '../messages';
// components
import Loading from '../components/UI/Loading/Loading';
import Layout from '../hoc/Layout';
import Title from '../components/Page/Title';
import CardsContainer from '../components/Page/CardsContainer';
import PageContainer from '../components/Page/PageContainer';
// selectors from reselect
import {
  selectQuery,
  selectSearchResultsLength,
  selectTotalResources,
  selectSearchWordLength,
  selectResourcesLoaded,
  selectResourcesReducerLoading,
  selectGroupedFilteredAvailableResources,
} from '../store/selectors';

import { SEARCH } from '../messages';
import ResourcePreview from '../components/ResourcePreview/ResourcePreview';

export class Component extends PureComponent {
  componentDidMount() {
    // flatted nodes from graphql
    if (!this.props.resourcesLoaded) {
      const resources = flattenGatsbyGraphQL(this.props.data.allDevhubSiphon.edges);
      this.props.loadResources(resources);
    }
  }

  componentDidUpdate() {
    const query = queryString.parse(this.props.location.search);
    if (Object.prototype.hasOwnProperty.call(query, 'q')) {
      const param = decodeURIComponent(query.q);

      if (param !== this.props.query) {
        this.props.setSearchQuery(param);
        this.getSearchResults(param).then(results => {
          this.props.setSearchResults(results);
        });
      }
    }
  }

  componentWillUnmount() {
    // unset set all search properties so that when this page is navigated back to, it looks like a fresh
    // page
    this.props.resetSearch();
  }

  /**
   * gets search results from lunr
   * @param {String} query the search string
   */
  async getSearchResults(query) {
    const lunr = await window.__LUNR__.__loaded;
    const lunrIndex = lunr.en;
    let results = [];
    // attempt to search by parsing query into fields
    try {
      results = lunrIndex.index.search(query);
    } catch (e) {
      // if that fails treat query as plain text and attempt search again
      results = lunrIndex.index.query(function() {
        this.term(query);
      });
    }
    // search results is an array of reference keys
    // we need to map those to the index store to get the actual
    // node ids
    const searchResultsMap = results
      .map(({ ref }) => lunrIndex.store[ref])
      .reduce((obj, result) => {
        obj[result.id] = { ...result };
        return obj;
      }, {});

    return searchResultsMap;
  }

  render() {
    const {
      resourcesByType,
      searchResultsLength,
      totalResources,
      setSearchBarTerms,
      searchWordLength,
      query,
    } = this.props;

    const resources = resourcesByType[RESOURCE_TYPES.COMPONENTS].map(r => ({
      type: r.resource.type,
      title: r.unfurl.title,
      description: r.unfurl.description,
      image: r.unfurl.image,
      path: r.resource.path,
    }));

    return (
      <Layout showHamburger>
        <main role="main" style={{ fontSize: '15px' }}>
          <Title
            title={COMPONENTS.header.title.defaultMessage}
            subtitle={COMPONENTS.header.subtitle.defaultMessage}
          />
          <PageContainer>
            <div>Filter menu here</div>
            {this.props.loading ? (
              <Loading message="Loading..." />
            ) : searchResultsLength === 0 && searchWordLength > 0 ? (
              <p>{SEARCH.results.empty.defaultMessage}</p>
            ) : (
              <CardsContainer resources={resources} setSearchBarTerms={setSearchBarTerms} />
            )}
          </PageContainer>
        </main>
      </Layout>
    );
  }
}

export const resourceQuery2 = graphql`
  query resourceQuery2 {
    allDevhubSiphon {
      edges {
        node {
          id
          name
          owner
          parent {
            id
          }
          _metadata {
            position
          }
          attributes {
            personas
          }
          source {
            displayName
            sourcePath
            type
            name
          }
          resource {
            path
            type
          }
          unfurl {
            title
            description
            type
            image
            author
          }
          childMarkdownRemark {
            frontmatter {
              pageOnly
            }
          }
        }
      }
    }
  }
`;

const mapStateToProps = createStructuredSelector({
  resourcesLoaded: selectResourcesLoaded,
  query: selectQuery,
  loading: selectResourcesReducerLoading,
  searchResultsLength: selectSearchResultsLength,
  totalResources: selectTotalResources,
  searchWordLength: selectSearchWordLength,
  resourcesByType: selectGroupedFilteredAvailableResources,
});

const mapDispatchToProps = dispatch => {
  return {
    loadResources: resources => dispatch(actions.loadResources(resources)),
    setSearchResults: results => dispatch(actions.setSearchResults(results)),
    setSearchQuery: query => dispatch(actions.setSearchQuery(query)),
    setSearchBarTerms: resourceType => dispatch(actions.setSearchBarTerms(resourceType)),
    resetSearch: () => dispatch(actions.resetSearch()),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Component);
