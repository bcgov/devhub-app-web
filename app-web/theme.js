/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING } from './src/constants/designTokens';
import { RESOURCE_TYPES, TOPICS, SEARCH_RESOURCE_TYPES, JOURNEY } from './src/constants/ui';

export const DEVHUB_PALETTE = {
  [RESOURCE_TYPES.COMPONENTS]: '#4299D2',
  [RESOURCE_TYPES.DOCUMENTATION]: '#246BD1',
  [RESOURCE_TYPES.PEOPLE]: '#E17039',
  [RESOURCE_TYPES.SELF_SERVICE_TOOLS]: '#F3BA45',
  [TOPICS]: '#444',
  [RESOURCE_TYPES.REPOSITORIES]: '#6e5494',
  [SEARCH_RESOURCE_TYPES.GITHUB_ISSUE]: '#6e5494',
  [RESOURCE_TYPES.EVENTS]: '#f44b42',
  [JOURNEY]: '#009688',
  lightgreen: '#66bb6a',
  yellow: '#FCBA19',
  lightyellow: '#F3BA45',
  grey: '#494949',
  lightgrey: '#F2F2F2',
  darkgrey: '#444',
  orange: '#E17039',
  purple: '#6E5494',
  red: '#D8292F',
  lightred: '#F44B42',
  green: '#2E8540',
  blue: '#003366',
  babyblue: '#4299D2',
  lightblue: '#246BD1',
  linkblue: '#1A5A96',
  bgBlue: '#38598A',
  white: '#ffffff',
};

export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
