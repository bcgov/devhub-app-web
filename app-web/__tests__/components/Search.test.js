import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import { Search, TEST_IDS } from '../../src/components/Search';
import { ThemeProvider } from 'emotion-theming';
import theme from '../../theme';

describe('Search Bar', () => {
  const keyupHandled = jest.fn();
  const search = jest.fn();
  const terms = '';
  const sampleText = 'foo';
  const inputConfig = {
    name: 'bar',
    id: 'baz',
  };

  beforeEach(() => {
    search.mockClear();
  });
  afterEach(cleanup);

  it('matches snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Search onkeyup={keyupHandled} onSearch={search} terms={terms} inputConfig={inputConfig} />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls the search handler when button is clicked and terms have length', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Search onkeyup={keyupHandled} onSearch={search} terms={terms} inputConfig={inputConfig} />
      </ThemeProvider>,
    );

    const button = queryByTestId(TEST_IDS.button);
    fireEvent.click(button);
    expect(search).toHaveBeenCalled();
  });

  it('doesn not call the search handle if the enter key is pressed with searchOnEnter = false', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Search
          onkeyup={keyupHandled}
          onSearch={search}
          terms={terms}
          inputConfig={inputConfig}
          searchOnEnter={false}
        />
      </ThemeProvider>,
    );

    let input = queryByTestId(TEST_IDS.input);
    // change text of input field
    fireEvent.change(input, { target: { value: sampleText } });
    // hit enter key
    // setting charcode to address known issue where keyPress event is not actually being fired
    // https://github.com/testing-library/react-testing-library/issues/269#issuecomment-455854112
    fireEvent.keyPress(input, { key: 'Enter', code: 13, charCode: 13 });

    expect(search).not.toHaveBeenCalled();
  });

  it(`calls the search handler with ${sampleText} when that text is entered and the enter key is pressed`, () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Search
          onkeyup={keyupHandled}
          onSearch={search}
          terms={terms}
          inputConfig={inputConfig}
          searchOnEnter={true}
        />
      </ThemeProvider>,
    );

    let input = queryByTestId(TEST_IDS.input);
    // change text of input field
    fireEvent.change(input, { target: { value: sampleText } });
    // hit enter key
    // setting charcode to address known issue where keyPress event is not actually being fired
    // https://github.com/testing-library/react-testing-library/issues/269#issuecomment-455854112
    fireEvent.keyPress(input, { key: 'Enter', code: 13, charCode: 13 });

    expect(search).toHaveBeenCalledWith(sampleText);
  });
});
