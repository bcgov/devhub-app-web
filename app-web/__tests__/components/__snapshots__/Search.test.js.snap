// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Bar matches snapshot 1`] = `
<Container
  onSearch={[MockFunction]}
  onSearchClear={[MockFunction]}
  onkeyup={[MockFunction]}
  searchOnEnter={false}
  terms=""
>
  <Input
    aria-label="Search input field"
    label={null}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    type="text"
    value=""
  />
  <Button
    aria-label="Click to Search"
    clicked={[Function]}
    type="primary"
  >
    Search
  </Button>
</Container>
`;
